### DISCUSSION TITLE
@ConfigProperty injection in custom Validator

### DISCUSSION THREAD

**Discussion by gquintana** [quarkusio/quarkus#36125] [Q&A] [answered]
I have a a custom validator which looks like:
```java
@Dependent
public class GreetingListValidator implements ConstraintValidator<GreetingList, String> {
    @ConfigProperty(name = "greeting.list", defaultValue = "hello")
    String greetingList;

    private Set<String> greetings = Collections.emptySet();

    @Override
    public void initialize(GreetingList constraintAnnotation) {
        System.out.println("Greeting List is " + greetingList+" in GreetingListValidator");
        greetings = Arrays.stream(greetingList.split(","))
                .map(String::trim).map(String::toLowerCase)
                .collect(Collectors.toSet());
    }

    @Override
    public boolean isValid(String value, ConstraintValidatorContext context) {
        return greetings.contains(value.trim().toLowerCase());
    }
}
```
I compile (with native compilation) and deploy my app.
At deploy time, I customize the `greeting.list` config with an environment variable:
```
GREETING_LIST=hello,bonjour
```
The effective greeting list is the default one: _hello_ . It does not contain _bonjour_. The external configuration is not taken. :'(
The `greetingList` field is injected (it's not null), but it contains the default value, not the externaly provided one.
I can not reproduce the problem in Quarkus Dev Mode, I need to go through the whole compilation phase.

Now, as a workaround, I moved my validation code into a CDI bean:
```java
@ApplicationScoped
public class GreetingListService {
    @ConfigProperty(name = "greeting.list", defaultValue = "hello")
    String greetingList;

    private Set<String> greetings = Collections.emptySet();

    @PostConstruct
    public void initialize() {
        System.out.println("Greeting List is " + greetingList+" in GreetingListValidator");
        greetings = Arrays.stream(greetingList.split(","))
                .map(String::trim).map(String::toLowerCase)
                .collect(Collectors.toSet());
    }

    public boolean isValid(String value) {
        return greetings.contains(value.trim().toLowerCase());
    }
}
@Dependent
public class GreetingListValidator implements ConstraintValidator<GreetingList, String> {
    @Inject
    String greetingListService;

    @Override
    public void initialize(GreetingList constraintAnnotation) {
    }

    @Override
    public boolean isValid(String value, ConstraintValidatorContext context) {
        return greetingListService.isValid(value);
    }
}
```
And this works! The environment variable is taken. :-)

My question is about Validator lifecycle. Are they created as CDI beans? What is the difference between a Validator initialization and a CDI bean inialization regarding @ConfigProperty injection

[a comment by an automated account is not shown]

**Comment by gquintana** [comment]
Answering to myself, I have just noticed during native compilation:

```
========================================================================================================================
GraalVM Native Image: Generating 'test-rest-validator-1.0.0-SNAPSHOT-runner' (executable)...
========================================================================================================================
For detailed information and explanations on the build output, visit:
https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md
------------------------------------------------------------------------------------------------------------------------
[1/8] Initializing...                                                                                    (8.9s @ 0.16GB)
 Java version: 17.0.8+7, vendor version: Mandrel-23.0.1.2-Final
 Graal compiler: optimization level: 2, target machine: x86-64-v3
 C compiler: gcc (redhat, x86_64, 8.5.0)
 Garbage collector: Serial GC (max heap size: 80% of RAM)
 4 user-specific feature(s)
 - io.quarkus.hibernate.validator.runtime.DisableLoggingFeature: Disables INFO logging during the analysis phase for the [org.hibernate.validator.internal.util.Version] categories
 - io.quarkus.runner.Feature: Auto-generated class by Quarkus from the existing extensions
 - io.quarkus.runtime.graal.DisableLoggingFeature: Disables INFO logging during the analysis phase
 - org.eclipse.angus.activation.nativeimage.AngusActivationFeature
Greeting List is hello in GreetingListValidator
[2/8] Performing analysis...  [******]                                                                  (51.3s @ 1.61GB)
```
The `@ConfigProperty` is injected at build time, not a initialization time

**Comment by mkouba** [comment, marked as answer]
> The `@ConfigProperty` is injected at build time, not a initialization time

Well, not exactly at build time. The CDI container is started during the `STATIC_INIT` [bootstrap phase](https://quarkus.io/guides/writing-extensions#bootstrap-three-phases). For a native executable build, however, the `STATIC_INIT` bootstrap phase runs as part of the native build process. I.e. if a CDI bean is initialized during `STATIC_INIT`  of a native executable build then the `@ConfigProperty` injection points cannot reflect the runtime values (configured with system properteis and ENV variables). And it seems that `ConstraintValidator`s are initialized during the `STATIC_INIT` phase.

**Comment by gquintana** [comment]
@mkouba Thanks for explaining, I am not sure to grasp everything though, it seems a bit complicated.

Lesson learned, I tried several alternatives.
First, I wrapped the config in Provider, to delay as much as possible config injection
```java
@Dependent
public class GreetingListValidator implements ConstraintValidator<GreetingList, String> {
    @ConfigProperty(name = "greeting.list", defaultValue = "hello")
    Provider<String> greetingList;

    private Set<String> greetings;

    @Override
    public void initialize(GreetingList constraintAnnotation) {
          String s = greetingList.get();
          System.out.println("Greeting List is " + s + " in GreetingListValidator");
          greetings = Arrays.stream(s.split(","))
                  .map(String::trim).map(String::toLowerCase)
                  .collect(Collectors.toSet());
    }

    @Override
    public boolean isValid(String value, ConstraintValidatorContext context) {
        return greetings.contains(value.trim().toLowerCase());
    }
}
```
Yet this is not enough. The `initialize` method still gets called during native build.
So I moved the `greetings` initialization later in the `isValid` method

```java
@Dependent
public class GreetingListValidator implements ConstraintValidator<GreetingList, String> {
    @ConfigProperty(name = "greeting.list", defaultValue = "hello")
    Provider<String> greetingList;

    private Set<String> greetings;

    @Override
    public void initialize(GreetingList constraintAnnotation) {
    }

    @Override
    public boolean isValid(String value, ConstraintValidatorContext context) {
        if (greetings == null) {
            String s = greetingList.get();
            System.out.println("Greeting List is " + s + " in GreetingListValidator");
            greetings = Arrays.stream(s.split(","))
                    .map(String::trim).map(String::toLowerCase)
                    .collect(Collectors.toSet());
        }
        return greetings.contains(value.trim().toLowerCase());
    }
}
```
This another possible workaround.
Is it possible to force the `initialize` method to me executed in the runtime init instead of static init?

**Reply by mkouba** [reply]
> Is it possible to force the initialize method to me executed in the runtime init instead of static init?

That's more a question for Bean Validation guys :-) 

CC @gsmet @yrodiere

**Reply by yrodiere** [reply]
I'm really not a "Bean Validation guy", but seeing Guillaume is on PTO...

No, I don't know of a BV-specific way to initialize stuff on runtime init.

I'd look into the Quarkus-specific startup initialization features: https://quarkus.io/guides/cdi-reference#startup_event

You'd also need a way to retrieve runtime config, so not using annotations... I suppose if you use some Smallrye API, you can achieve that?

**Reply by yrodiere** [reply]
I misread the initial question, so I'll rephrase my answer:

No, I don't know of a BV-specific way to initialize stuff on runtime init.

Your solution of injecting a `Provider<String>` is the right one, though you could use Quarkus-specific startup initialization features to avoid the lazy initialization: https://quarkus.io/guides/cdi-reference#startup_event

**Comment by gquintana** [comment]
I am still puzzled by the fact that validator init doesn't behave like normal CDI bean init.
Having spent more than a day debugging this issue before pinpointing that the problem was related to native compilation, I can tell this behaviour is not intuitive to me. The native application should behave like the normal JVM based application.

**Reply by yrodiere** [reply]
It's a fundamental design choice in Quarkus: static init (and a whole lot of initialization) happens at build time for native applications (as evidenced [here](https://quarkus.io/guides/lifecycle#what-is-the-difference-from-initializedapplicationscoped-class-and-destroyedapplicationscoped-class)), the main reason being that such early init allows much better dead code elimination (smaller executable size and memory footprint) as well as faster startup. Which are what you're after when compiling to native.

You may disagree with that choice, but I personally am unable to think of a solution that would bring the performance benefits of this design without the different behavior.

**Reply by yrodiere** [reply]
Or rather, I can think of such a solution, and it might happen (or not) eventually with all the buzz about CRIU lately: moving static init to build time even for JVM applications.

Probably not what you want, though :)

**Reply by mkouba** [reply]
Hm, pls define the "normal CDI bean init". I can't comment on why the `ConstraintValidator`s are initialized during STATIC_INIT but other than that the beans are initialized in a standard way.

**Reply by gquintana** [reply]
I mean in a "normal CDI bean init", a field annotated @ConfigProperty get the value when the application starts,
whereas in a validator, a field annotated @ConfigProperty get the value when the application is natively built.

**Reply by yrodiere** [reply]
Beans are initialized when they are first needed, IIRC.

Hibernate Validator starts during static init (for the reasons mentioned above), so the beans it depends on (like `ConstraintValidator`s) are initialized then, and take the value of the configuration property during static init.

It might be possible to delay the call to the `init()` method to runtime init, but I don't think we can delay the retrieval of `ConstraintValidator` beans, because I think Hibernate Validator calls some methods and uses some reflection on those beans to configure itself. @gsmet would know better than I.

I agree injecting beans with build-time config can be surprising though, especially in situations like this where the bean lifecycle is not controlled by application developers. @mkouba @radcortez Shouldn't we throw an exception by default when attempting to inject a `@ConfigProperty` into a bean at build-time, and require build-time config injection to be explicit (e.g. by requiring an annotation like `@BuildConfig` or whatever)? Of course it's a breaking change and would require mitigation strategies, but apart from that, do you think this would make sense / be helpful to developers?

**Reply by gquintana** [reply]
It's crystal clear, thanks for taking time to explain.

**Reply by mkouba** [reply]
> Shouldn't we throw an exception by default when attempting to inject a `@ConfigProperty` into a bean at build-time, and require build-time config injection to be explicit (e.g. by requiring an annotation like `@BuildConfig` or whatever)? Of course it's a breaking change and would require mitigation strategies, but apart from that, do you think this would make sense / be helpful to developers?

I'm not sure about throwing an exception but a warning with an actionable message could be helpful.

And it seems that we already have `ConfigInjectionStaticInitBuildItem` - a mechanism to detect injected fields into a class annotated with a `@Provider` and `@WebFilter`. We could extend this mechanism to cover `ConstraintValidator`s as well. But the warning recommends the usage of `@Inject @ConfigProperty(name = "foo") Instance<String>` which would not help in this particular case...

Moreover, it's perfectly fine to inject a config property that cannot be changed at runtime (e.g. a quarkus config item with `ConfigPhase.BUILD_AND_RUN_TIME_FIXED`). So it's not 100% reliable :shrug:.

**Reply by yrodiere** [reply]
> I'm not sure about throwing an exception but a warning with an actionable message could be helpful.

Gavin would tell you warnings are useless because they are too easily missed, and I tend to agree. An exception by default with a configuration knob to disable the check makes more sense IMO.

> But the warning recommends the usage of `@Inject @ConfigProperty(name = "foo") Instance<String>` which would not help in this particular case...

Well that's more or less the approach @gquintana took [in the end](https://github.com/quarkusio/quarkus/discussions/36125#discussioncomment-7100949), though you're right that they had to also move the config property retrieval to another method.

> Moreover, it's perfectly fine to inject a config property that cannot be changed at runtime (e.g. a quarkus config item with `ConfigPhase.BUILD_AND_RUN_TIME_FIXED`). So it's not 100% reliable 🤷.

It's certainly fine, and if we know the config property is fixed at build time we might want to skip the warning/error. For user properties though, we have no such knowledge: they could be runtime or build time, we have no idea.

My point was that if a (potentially runtime) config property is injected into a user bean, and that bean is initialized at static init, the situation is exceptional enough that we might want people to be explicit about it. Throw an exception and ask them to confirm they are fine with that configuration property being fixed at static init, for example with an annotation on the config property being injected.

**Reply by mkouba** [reply]
> > But the warning recommends the usage of `@Inject @ConfigProperty(name = "foo") Instance<String>` which would not help in this particular case...
> 
> Well that's more or less the approach @gquintana took [in the end](https://github.com/quarkusio/quarkus/discussions/36125#discussioncomment-7100949), though you're right that they had to also move the config property retrieval to another method.

Yes, because if an extension also calls a method upon the component then lazy instantiation does not help at all.

> My point was that if a (potentially runtime) config property is injected into a user bean, and that bean is initialized at static init, the situation is exceptional enough that we might want people to be explicit about it. Throw an exception and ask them to confirm they are fine with that configuration property being fixed at static init, for example with an annotation on the config property being injected.

Well, I can imagine that we fail the build if we're pretty sure that it's a real problem. Otherwise we'll just annoy the users with yet another configuration knob.

So we could throw an exception:

1. If we find a `@ConfigProperty` injection point (field, constructor, initializer) declared on a bean that represents a component of an extension that is initialized during STATIC_INIT (e.g. `ConstraintValidator` or `@Provider`)
2. The extension should indicate if the component is just collected (e.g. client proxy and `@ApplicationScoped` might help) or whether it's actually used (e.g. `ConstraintValidator.initialize()` is called; here client proxy and even `Instance<>` might not be enough). 
3. `BUILD_AND_RUN_TIME_FIXED` properties should be ignored
 
The question is whether we should only validate the injection points when the native image is being built. I guess that **we should** because that's the only situation where it matters.

Like you mentioned we should also provide a way to eliminate false positives. An annotation with a clear name should be fine. But naming is hard, right ;-). Maybe something like `@NativeBuildFixed`...

**Reply by yrodiere** [reply]
> If we find a @ConfigProperty injection point (field, constructor, initializer) declared on a bean that represents a component of an extension that is initialized during STATIC_INIT (e.g. ConstraintValidator or @Provider)

Not sure what "represents a component of an extension [...]" means. Just in case you mean "something declared explicitly by the extension as problematic", I'd like to mention that this problem could affect potentially any bean, e.g. a completely custom bean (no implemented interface) injected into a `ConstraintValidator`.

That's one reason why I was thinking of a static init validation rather than build time validation (another reason below).

> even `Instance<>` might not be enough).

Probably unrelated to your point, but I was thinking that for injected `Instance`, you might want to delay the check until the property is actually retrieved. If it's injected at static init but retrieved at runtime init, then we're fine.

So it's not about validating injection points, but more about validating... injections? IDK the proper wording.

> The question is whether we should only validate the injection points when the native image is being built. I guess that we should because that's the only situation where it matters.

Makes sense to me.

> Like you mentioned we should also provide a way to eliminate false positives. An annotation with a clear name should be fine. But naming is hard, right ;-). Maybe something like `@NativeBuildFixed`...

`@NativeBuildTime`?

Anyway, that's something we could happily bikeshed on once we open an issue/PR :) If you think it makes sense to open one.

**Reply by gquintana** [reply]
We also noticed the `@ConfigProperty` behaved differently dependending if the ConstraintValidator was annotated `@ApplicationScoped` or `@Dependent`. Obviously the validator lifecycle is different, and it makes sense. But for the newcomer, it's surprising.

**Reply by mkouba** [reply]
> So it's not about validating injection points, but more about validating... injections? IDK the proper wording.

Yes, we could validate the `ImageMode` when an injected `@ConfigProperty` is being created. The downside I can spot immediately is that a thrown exception doesn't necessarily have to result in a static init failure (for example an extension can swallow the exception when obtaining the bean with `Instance`).

> Anyway, that's something we could happily bikeshed on once we open an issue/PR :) If you think it makes sense to open one.

@yrodiere Yes, I think that it makes sense and I will try to prepare a prototype that could be discussed further...

> We also noticed the @ConfigProperty behaved differently dependending if the ConstraintValidator was annotated @ApplicationScoped or @Dependent. Obviously the validator lifecycle is different, and it makes sense. But for the newcomer, it's surprising.

@gquintana Yes, a `@Dependent` bean is created immediately when injected/obtained programmatically. Whereas `@ApplicationScoped` is created lazily, when a method is called for the first time (see https://quarkus.io/guides/cdi-reference#lazy_by_default for more info). I understand that it's a bit surprising. On the other hand, it's well-defined and documented.

**Reply by yrodiere** [reply]
> The downside I can spot immediately is that a thrown exception doesn't necessarily have to result in a static init failure (for example an extension can swallow the exception when obtaining the bean with Instance).

`throw new Throwable();` :imp: 

More seriously, you could also store such exceptions in a collection that would be checked after static init, and make sure to report them *then* (potentially as suppressed exception, if static init failed for another reason). Not perfect, but probably the best we can do.

> @yrodiere Yes, I think that it makes sense and I will try to prepare a prototype that could be discussed further...

Thanks a lot @mkouba!
